File: apps/dsa-web/src/api/market.ts
import apiClient from './index';

export type NewHighDateItem = {
  date: string;
  hfq_close: number;
};

export type HfqNewHighItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close: number;
  new_high_count: number;
  current_hfq_close?: number | null;
  drawdown_from_high_pct?: number | null;
  ytd_hfq_return_pct?: number | null;
  new_high_dates: NewHighDateItem[];
};

export type HfqNewHighListResponse = {
  start_date: string;
  as_of_date: string;
  total: number;
  items: HfqNewHighItem[];
};

export type HfqBollPickItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close?: number | null;
  current_hfq_close: number;
  drawdown_from_high_pct?: number | null;
  boll_mid: number;
  boll_lower: number;
  boll_upper?: number;
  dist_mid_pct: number;
  dist_lower_pct: number;
  dist_upper_pct?: number;
  band_zone: string;
};

export type HfqBollPickListResponse = {
  start_date: string;
  as_of_date: string;
  lookback_days: number;
  near_pct: number;
  max_drawdown_from_high_pct: number;
  cutoff_date: string;
  total: number;
  items: HfqBollPickItem[];
};

export type HfqKLineItem = {
  date: string;
  open?: number | null;
  high?: number | null;
  low?: number | null;
  close: number;
  volume?: number | null;
};

export type HfqKLineResponse = {
  stock_code: string;
  start_date: string;
  end_date: string;
  data: HfqKLineItem[];
};

// ── ETF 新高 ──

export type EtfNewHighDateItem = {
  date: string;
  close: number;
};

export type EtfNewHighItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close: number;
  new_high_count: number;
  current_close?: number | null;
  drawdown_from_high_pct?: number | null;
  ytd_return_pct?: number | null;
  new_high_dates: EtfNewHighDateItem[];
};

export type EtfNewHighListResponse = {
  start_date: string;
  as_of_date: string;
  total: number;
  items: EtfNewHighItem[];
};

export type EtfBollPickItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close?: number | null;
  current_close: number;
  drawdown_from_high_pct?: number | null;
  boll_mid: number;
  boll_lower: number;
  boll_upper?: number;
  dist_mid_pct: number;
  dist_lower_pct: number;
  dist_upper_pct?: number;
  band_zone: string;
  mid_slope?: number | null;
};

export type EtfBollPickListResponse = {
  start_date: string;
  as_of_date: string;
  lookback_days: number;
  near_pct: number;
  max_drawdown_from_high_pct: number;
  cutoff_date: string;
  total: number;
  items: EtfBollPickItem[];
};

export type EtfKLineResponse = {
  stock_code: string;
  start_date: string;
  end_date: string;
  data: HfqKLineItem[];
};

// ── 全球主要指数 ──

export type GlobalIndexNewHighDateItem = {
  date: string;
  close: number;
};

export type GlobalIndexNewHighItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close: number;
  new_high_count: number;
  current_close?: number | null;
  drawdown_from_high_pct?: number | null;
  ytd_return_pct?: number | null;
  new_high_dates: GlobalIndexNewHighDateItem[];
};

export type GlobalIndexNewHighListResponse = {
  start_date: string;
  as_of_date: string;
  total: number;
  items: GlobalIndexNewHighItem[];
};

export type GlobalIndexKLineResponse = {
  ts_code: string;
  start_date: string;
  end_date: string;
  data: HfqKLineItem[];
};

export type GlobalIndexBollPickListResponse = {
  start_date: string;
  as_of_date: string;
  lookback_days: number;
  near_pct: number;
  max_drawdown_from_high_pct: number;
  cutoff_date: string;
  total: number;
  items: Array<{
    ts_code: string;
    stock_code: string;
    stock_name: string;
    latest_new_high_date: string;
    latest_new_high_close: number | null;
    current_close: number;
    drawdown_from_high_pct: number | null;
    boll_mid: number;
    boll_lower: number;
    boll_upper: number | null;
    dist_mid_pct: number;
    dist_lower_pct: number;
    dist_upper_pct: number | null;
    band_zone: string;
  }>;
};

// ── A 股指数 ──

export type AIndexNewHighDateItem = {
  date: string;
  close: number;
};

export type AIndexNewHighItem = {
  ts_code: string;
  stock_code: string;
  stock_name: string;
  latest_new_high_date: string;
  latest_new_high_close: number;
  new_high_count: number;
  current_close?: number | null;
  drawdown_from_high_pct?: number | null;
  ytd_return_pct?: number | null;
  new_high_dates: AIndexNewHighDateItem[];
};

export type AIndexNewHighListResponse = {
  start_date: string;
  as_of_date: string;
  total: number;
  items: AIndexNewHighItem[];
};

export type AIndexKLineResponse = {
  ts_code: string;
  start_date: string;
  end_date: string;
  freq: string;
  data: HfqKLineItem[];
};

export type AIndexBollPickListResponse = {
  start_date: string;
  as_of_date: string;
  lookback_days: number;
  near_pct: number;
  max_drawdown_from_high_pct: number;
  cutoff_date: string;
  total: number;
  items: Array<{
    ts_code: string;
    stock_code: string;
    stock_name: string;
    latest_new_high_date: string;
    latest_new_high_close: number | null;
    current_close: number;
    drawdown_from_high_pct: number | null;
    boll_mid: number;
    boll_lower: number;
    boll_upper: number | null;
    dist_mid_pct: number;
    dist_lower_pct: number;
    dist_upper_pct: number | null;
    band_zone: string;
  }>;
};

export const marketApi = {
  async getHfqNewHighs(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean }): Promise<HfqNewHighListResponse> {
    const response = await apiClient.get<HfqNewHighListResponse>('/api/v1/market/hfq-new-highs', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getHfqBollPicks(opts?: {
    startDate?: string;
    asOfDate?: string;
    refresh?: boolean;
    nearPct?: number;
    lookbackDays?: number;
    maxDrawdownFromHighPct?: number;
  }): Promise<HfqBollPickListResponse> {
    const response = await apiClient.get<HfqBollPickListResponse>('/api/v1/market/hfq-new-highs/boll-picks', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
        near_pct: opts?.nearPct ?? 2,
        lookback_days: opts?.lookbackDays ?? 30,
        max_drawdown_from_high_pct: opts?.maxDrawdownFromHighPct ?? 20,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getHfqKlines(stockCode: string, startDate = '20260101', endDate?: string): Promise<HfqKLineResponse> {
    const response = await apiClient.get<HfqKLineResponse>(`/api/v1/market/hfq-new-highs/${encodeURIComponent(stockCode)}/klines`, {
      params: {
        start_date: startDate,
        end_date: endDate,
      },
      timeout: 30000,
    });
    return response.data;
  },

  // ── ETF 新高 ──

  async getEtfNewHighs(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean }): Promise<EtfNewHighListResponse> {
    const response = await apiClient.get<EtfNewHighListResponse>('/api/v1/market/etf-new-highs', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getEtfBollPicks(opts?: {
    startDate?: string;
    asOfDate?: string;
    refresh?: boolean;
    nearPct?: number;
    lookbackDays?: number;
    maxDrawdownFromHighPct?: number;
  }): Promise<EtfBollPickListResponse> {
    const response = await apiClient.get<EtfBollPickListResponse>('/api/v1/market/etf-new-highs/boll-picks', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
        near_pct: opts?.nearPct ?? 2,
        lookback_days: opts?.lookbackDays ?? 30,
        max_drawdown_from_high_pct: opts?.maxDrawdownFromHighPct ?? 30,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getEtfKlines(etfCode: string, startDate = '20260101', endDate?: string): Promise<EtfKLineResponse> {
    const response = await apiClient.get<EtfKLineResponse>(`/api/v1/market/etf-new-highs/${encodeURIComponent(etfCode)}/klines`, {
      params: {
        start_date: startDate,
        end_date: endDate,
      },
      timeout: 30000,
    });
    return response.data;
  },

  // ── 全球主要指数 ──

  async getGlobalIndexNewHighs(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean }): Promise<GlobalIndexNewHighListResponse> {
    const response = await apiClient.get<GlobalIndexNewHighListResponse>('/api/v1/market/global-index-new-highs', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getGlobalIndexBollPicks(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean }): Promise<GlobalIndexBollPickListResponse> {
    const response = await apiClient.get<GlobalIndexBollPickListResponse>('/api/v1/market/global-index-new-highs/boll-picks', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getGlobalIndexKlines(tsCode: string, startDate = '20260101', endDate?: string): Promise<GlobalIndexKLineResponse> {
    const response = await apiClient.get<GlobalIndexKLineResponse>(`/api/v1/market/global-index-new-highs/${encodeURIComponent(tsCode)}/klines`, {
      params: {
        start_date: startDate,
        end_date: endDate,
      },
      timeout: 30000,
    });
    return response.data;
  },

  // ── A 股指数（日线/周线）──

  async getAIndexList(): Promise<{ total: number; items: Array<{ ts_code: string; name: string }> }> {
    const response = await apiClient.get('/api/v1/market/a-index-list', { timeout: 30000 });
    return response.data;
  },

  async getAIndexNewHighs(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean; freq?: string }): Promise<AIndexNewHighListResponse> {
    const response = await apiClient.get<AIndexNewHighListResponse>('/api/v1/market/a-index-new-highs', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
        freq: opts?.freq ?? 'daily',
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getAIndexBollPicks(opts?: { startDate?: string; asOfDate?: string; refresh?: boolean; nearPct?: number; lookbackDays?: number; maxDrawdownFromHighPct?: number; freq?: string }): Promise<AIndexBollPickListResponse> {
    const response = await apiClient.get<AIndexBollPickListResponse>('/api/v1/market/a-index-new-highs/boll-picks', {
      params: {
        start_date: opts?.startDate ?? '20260101',
        as_of_date: opts?.asOfDate,
        refresh: opts?.refresh ?? false,
        near_pct: opts?.nearPct ?? 2,
        lookback_days: opts?.lookbackDays ?? 30,
        max_drawdown_from_high_pct: opts?.maxDrawdownFromHighPct ?? 30,
        freq: opts?.freq ?? 'daily',
      },
      timeout: 300000,
    });
    return response.data;
  },

  async getAIndexKlines(tsCode: string, startDate = '20260101', endDate?: string, freq = 'daily'): Promise<AIndexKLineResponse> {
    const response = await apiClient.get<AIndexKLineResponse>(`/api/v1/market/a-index-new-highs/${encodeURIComponent(tsCode)}/klines`, {
      params: { start_date: startDate, end_date: endDate, freq },
      timeout: 30000,
    });
    return response.data;
  },

  async getAIndexConstituents(indexCode: string): Promise<{ index_code: string; total: number; items: Array<{ con_code: string; con_name: string | null; weight: number | null }> }> {
    const response = await apiClient.get(`/api/v1/market/a-index-constituents/${encodeURIComponent(indexCode)}`, {
      timeout: 30000,
    });
    return response.data;
  },
};

File: api/v1/schemas/market.py
# -*- coding: utf-8 -*-
"""Market statistics API schemas."""

from typing import List, Optional

from pydantic import BaseModel, Field


class NewHighDateItem(BaseModel):
    date: str = Field(..., description="创新高日期 YYYYMMDD")
    hfq_close: float = Field(..., description="后复权收盘价")


class HfqNewHighItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: float
    new_high_count: int
    current_hfq_close: Optional[float] = None
    drawdown_from_high_pct: Optional[float] = None
    ytd_hfq_return_pct: Optional[float] = Field(
        None, description="2026 年初首个交易日后复权收盘至今涨幅（%）"
    )
    new_high_dates: List[NewHighDateItem] = Field(default_factory=list)


class HfqNewHighListResponse(BaseModel):
    start_date: str
    as_of_date: str
    total: int
    items: List[HfqNewHighItem] = Field(default_factory=list)


class HfqBollPickItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: Optional[float] = None
    current_hfq_close: float
    drawdown_from_high_pct: Optional[float] = Field(None, description="现价相对最近新高偏离（%），越接近 0 越近新高")
    boll_mid: float
    boll_lower: float
    boll_upper: Optional[float] = None
    dist_mid_pct: float = Field(..., description="现价相对 BOLL 中轨偏离（%）")
    dist_lower_pct: float = Field(..., description="现价相对 BOLL 下轨偏离（%）")
    dist_upper_pct: Optional[float] = Field(None, description="现价相对 BOLL 上轨偏离（%）")
    band_zone: str = Field(..., description="mid / lower / upper / 组合如 mid_lower / upper_mid")


class HfqBollPickListResponse(BaseModel):
    start_date: str
    as_of_date: str
    lookback_days: int
    near_pct: float
    max_drawdown_from_high_pct: float = Field(
        default=20.0,
        description="距最近新高最大回撤（%），超出则排除",
    )
    cutoff_date: str
    total: int
    items: List[HfqBollPickItem] = Field(default_factory=list)


class HfqKLineItem(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: float
    volume: Optional[float] = None


class HfqKLineResponse(BaseModel):
    stock_code: str
    start_date: str
    end_date: str
    data: List[HfqKLineItem] = Field(default_factory=list)


class HkGgtComponentItem(BaseModel):
    trade_date: str
    hk_code: str
    name: Optional[str] = None
    latest_price: Optional[float] = None
    pct_change: Optional[float] = None
    change_amount: Optional[float] = None
    open: Optional[float] = Field(None, alias="open")
    high: Optional[float] = None
    low: Optional[float] = None
    prev_close: Optional[float] = None
    volume: Optional[float] = None
    amount: Optional[float] = None

    class Config:
        populate_by_name = True


class HkGgtComponentListResponse(BaseModel):
    trade_date: str
    total: int
    items: List[HkGgtComponentItem] = Field(default_factory=list)
    available_dates: List[str] = Field(default_factory=list)


class HkStockListItem(BaseModel):
    hk_code: str
    name: Optional[str] = None
    latest_price: Optional[float] = None
    pct_change: Optional[float] = None
    boll_mid: Optional[float] = None
    boll_upper: Optional[float] = None
    boll_lower: Optional[float] = None
    boll_mid_dist_pct: Optional[float] = None
    boll_upper_dist_pct: Optional[float] = None
    boll_lower_dist_pct: Optional[float] = None


class HkStockListResponse(BaseModel):
    trade_date: str
    total: int
    items: List[HkStockListItem] = Field(default_factory=list)


class HkStockKLineItem(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: float
    volume: Optional[float] = None
    boll_mid: Optional[float] = None
    boll_upper: Optional[float] = None
    boll_lower: Optional[float] = None


class HkStockKLineResponse(BaseModel):
    hk_code: str
    start_date: str
    end_date: str
    data: List[HkStockKLineItem] = Field(default_factory=list)


class HkBollPickItem(BaseModel):
    hk_code: str
    name: str
    close: float
    band: str
    boll_mid: float
    boll_upper: float
    boll_lower: float
    dist_pct: Optional[float] = None


class HkBollPickListResponse(BaseModel):
    near_pct: float
    upper: List[HkBollPickItem] = Field(default_factory=list)
    mid: List[HkBollPickItem] = Field(default_factory=list)
    lower: List[HkBollPickItem] = Field(default_factory=list)


class HkGgtPollResponse(BaseModel):
    polled: bool
    reason: Optional[str] = None
    trade_date: Optional[str] = None
    bar_time: Optional[str] = None
    saved: int = 0
    matched: Optional[int] = None


class EtfNewHighDateItem(BaseModel):
    date: str = Field(..., description="创新高日期 YYYYMMDD")
    close: float = Field(..., description="收盘价")


class EtfNewHighItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: float
    new_high_count: int
    current_close: Optional[float] = None
    drawdown_from_high_pct: Optional[float] = None
    ytd_return_pct: Optional[float] = Field(
        None, description="2026 年初首个交易日至今涨幅（%）"
    )
    new_high_dates: List[EtfNewHighDateItem] = Field(default_factory=list)


class EtfNewHighListResponse(BaseModel):
    start_date: str
    as_of_date: str
    total: int
    items: List[EtfNewHighItem] = Field(default_factory=list)


class EtfBollPickItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: Optional[float] = None
    current_close: float
    drawdown_from_high_pct: Optional[float] = Field(None, description="现价相对最近新高偏离（%），越接近 0 越近新高")
    boll_mid: float
    boll_lower: float
    boll_upper: Optional[float] = None
    dist_mid_pct: float = Field(..., description="现价相对 BOLL 中轨偏离（%）")
    dist_lower_pct: float = Field(..., description="现价相对 BOLL 下轨偏离（%）")
    dist_upper_pct: Optional[float] = Field(None, description="现价相对 BOLL 上轨偏离（%）")
    band_zone: str = Field(..., description="mid / lower / upper / 组合")
    mid_slope: Optional[float] = None


class EtfBollPickListResponse(BaseModel):
    start_date: str
    as_of_date: str
    lookback_days: int
    near_pct: float
    max_drawdown_from_high_pct: float = Field(
        default=30.0,
        description="距最近新高最大回撤（%），超出则排除",
    )
    cutoff_date: str
    total: int
    items: List[EtfBollPickItem] = Field(default_factory=list)


class EtfKLineItem(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: float
    volume: Optional[float] = None


class EtfKLineResponse(BaseModel):
    stock_code: str
    start_date: str
    end_date: str
    data: List[EtfKLineItem] = Field(default_factory=list)


class GlobalIndexNewHighDateItem(BaseModel):
    date: str = Field(..., description="创新高日期 YYYYMMDD")
    close: float = Field(..., description="收盘价")


class GlobalIndexNewHighItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: float
    new_high_count: int
    current_close: Optional[float] = None
    drawdown_from_high_pct: Optional[float] = None
    ytd_return_pct: Optional[float] = Field(
        None, description="2026 年初首个交易日至今涨幅（%）"
    )
    new_high_dates: List[GlobalIndexNewHighDateItem] = Field(default_factory=list)


class GlobalIndexNewHighListResponse(BaseModel):
    start_date: str
    as_of_date: str
    total: int
    items: List[GlobalIndexNewHighItem] = Field(default_factory=list)


class GlobalIndexKLineItem(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: float
    volume: Optional[float] = None


class GlobalIndexKLineResponse(BaseModel):
    ts_code: str
    start_date: str
    end_date: str
    data: List[GlobalIndexKLineItem] = Field(default_factory=list)


class GlobalIndexBollPickItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: Optional[float] = None
    current_close: float
    drawdown_from_high_pct: Optional[float] = None
    boll_mid: float
    boll_lower: float
    boll_upper: Optional[float] = None
    dist_mid_pct: float
    dist_lower_pct: float
    dist_upper_pct: Optional[float] = None
    band_zone: str


class GlobalIndexBollPickListResponse(BaseModel):
    start_date: str
    as_of_date: str
    lookback_days: int
    near_pct: float
    max_drawdown_from_high_pct: float = 30.0
    cutoff_date: str
    total: int
    items: List[GlobalIndexBollPickItem] = Field(default_factory=list)


# ── A 股指数（index_daily / index_weekly）──


class AIndexListItem(BaseModel):
    ts_code: str
    name: Optional[str] = None
    fullname: Optional[str] = None
    market: Optional[str] = None
    publisher: Optional[str] = None
    index_type: Optional[str] = None
    category: Optional[str] = None


class AIndexListResponse(BaseModel):
    total: int
    items: List[AIndexListItem] = Field(default_factory=list)


class AIndexNewHighDateItem(BaseModel):
    date: str = Field(..., description="创新高日期 YYYYMMDD")
    close: float = Field(..., description="收盘价")


class AIndexNewHighItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: float
    new_high_count: int
    current_close: Optional[float] = None
    drawdown_from_high_pct: Optional[float] = None
    ytd_return_pct: Optional[float] = None
    new_high_dates: List[AIndexNewHighDateItem] = Field(default_factory=list)


class AIndexNewHighListResponse(BaseModel):
    start_date: str
    as_of_date: str
    total: int
    items: List[AIndexNewHighItem] = Field(default_factory=list)


class AIndexKLineItem(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: float
    volume: Optional[float] = None


class AIndexKLineResponse(BaseModel):
    ts_code: str
    start_date: str
    end_date: str
    freq: str = "daily"
    data: List[AIndexKLineItem] = Field(default_factory=list)


class AIndexBollPickItem(BaseModel):
    ts_code: str
    stock_code: str
    stock_name: str
    latest_new_high_date: str
    latest_new_high_close: Optional[float] = None
    current_close: float
    drawdown_from_high_pct: Optional[float] = None
    boll_mid: float
    boll_lower: float
    boll_upper: Optional[float] = None
    dist_mid_pct: float
    dist_lower_pct: float
    dist_upper_pct: Optional[float] = None
    band_zone: str


class AIndexBollPickListResponse(BaseModel):
    start_date: str
    as_of_date: str
    lookback_days: int
    near_pct: float
    max_drawdown_from_high_pct: float = 30.0
    cutoff_date: str
    total: int
    items: List[AIndexBollPickItem] = Field(default_factory=list)


class AIndexConstituentItem(BaseModel):
    con_code: str
    con_name: Optional[str] = None
    weight: Optional[float] = None
    trade_date: Optional[str] = None


class AIndexConstituentResponse(BaseModel):
    index_code: str
    total: int
    items: List[AIndexConstituentItem] = Field(default_factory=list)
